# Numerical Optimization (CS215300) Assignment 3
## Introduction
In this assignment, we expect you to be able to solve constrained optimization problem by Scipy library. We want you to apply two algorithms on the following problem as benchmark, survey the method used in these libraries, and analysis the behaviour of these algorithms.
The library document link: https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html

## Task
1. Please solve the following problrem by using `trust-constr` method:

$$
        \begin{array}{cc}
                \displaystyle \min_{x_1,x_2} & f(x_1,x_2)=-x_1-x_2 \\
        \end{array}
$$

$$
        \text{s.t.} -2x_1^4 + 8x_1^3 -8x_1^2 + x_2 - 2 \le 0  \\
        -4x_1^4 + 32x_1^3 - 88x_1^2 + 96x_1 + x_2 -36 \le 0   \\
        0 \le x_1 \le 3 \\
        0 \le x_2 \le 4 \\
$$

2. Please use COBYLA method to solve the same problem.
3. In your report, please read the paper cited in the libraries, which gives the details of the algorithms. Write a brief introduction of the algorithms, and compare their behaviours in this benchmark. You are not necessarily to read the original paper, other resourses (ex. slides from other schools or surveys) are also acceptable, please include the link or paper name in your reference if you referred other resources.
4. Please provide the Jacobian and Hessian function in matrix form in your report. Basic latex syntax is supported in Markdown.
5. Rename this notebook file with your student ID and upload it to eeclass platform. (ex. 107xxxxxx.ipynb)

In [1]:
import numpy as np
from scipy.optimize import Bounds
from scipy.optimize import minimize
from scipy.optimize import NonlinearConstraint

### Define objective function

In [2]:
def f(x):
    return - x[0] - x[1]

def grad(x):
    return np.array([-1.0, -1.0])

def hess(x):
    return np.array([[0.0,0.0],[0.0,0.0]])

### Define constraint functions and derivatives

In [3]:
def cons_f(x):
    constrain1 = (-2.0*x[0]**4) +  (8.0*x[0]**3) -  (8.0*x[0]**2) + x[1] - 2.0
    constrain2 = (-4.0*x[0]**4) + (32.0*x[0]**3) - (88.0*x[0]**2) + (96.0*x[0]) + x[1] - 36.0
    return [constrain1, constrain2]

def cons_Jacobian(x):
    jacobian1 = [(-8.0*(x[0]**3)) + (24.0*(x[0]**2)) - (16.0*x[0]), 1.0]
    jacobian2 = [(-16.0*(x[0]**3)) + (96.0*(x[0]**2)) - (176.0*x[0]) + 96.0, 1.0]
    return [jacobian1, jacobian2]

def cons_Hessian(x, v):
    hessian1 = np.array([[(-24.0*x[0]**2) +  (48.0*x[0]) -  16.0, 0.0], [0.0, 0.0]])
    hessian2 = np.array([[(-48.0*x[0]**2) + (192.0*x[0]) - 176.0, 0.0], [0.0, 0.0]])
    return v[0]*hessian1 + v[1]*hessian2

nonlinear_constraint = NonlinearConstraint(
    cons_f,  
    -np.inf,
    0,
    jac = cons_Jacobian, 
    hess = cons_Hessian
)

### Call the minimize library

In [4]:
bounds = Bounds([0, 0], [3.0, 4.0])
x0 = [0,0]
res = minimize(
    f, 
    x0, 
    method = 'trust-constr', 
    jac = grad, 
    hess = hess, 
    constraints = [nonlinear_constraint], 
    options = {'verbose': 1},
    bounds = bounds
)
print(res.x)

`gtol` termination condition is satisfied.
Number of iterations: 17, function evaluations: 13, CG iterations: 12, optimality: 7.68e-09, constraint violation: 0.00e+00, execution time: 0.037 s.
[0.61157004 3.44188615]


### Apply COBYLA method

In [5]:
x0 = [0,0]
res = minimize(
    f, 
    x0, 
    method = 'cobyla', 
    constraints = [nonlinear_constraint]
)
print(res.x)

[0.61160344 3.44210482]


/Users/bijonsetyawan/.local/lib/python3.9/site-packages/scipy/optimize/_constraints.py:356: OptimizeWarning: Constraint options `finite_diff_jac_sparsity`, `finite_diff_rel_step`, `keep_feasible`, and `hess`are ignored by this method.
  warn("Constraint options `finite_diff_jac_sparsity`, "


# Report

In this homework, the result that I got from `trust-constr` and `cobyla` are exactly similar. The result is `[0.61160344 3.44210482]`.

## The Objective Function

$$
    f(x_1, x_2) = -x_1 - x_2
$$

## The Gradient Objective Function

$$
    \nabla f(x_1, x_2) = \begin{bmatrix} -1 & -1 \end{bmatrix}
$$

## The Hessian Objective Function

$$
    \nabla^2 f(x_1, x_2) = \begin{bmatrix} 0 & 0 \\ 0 & 0 \end{bmatrix}
$$

## The Constraints

$$
    \begin{bmatrix}
        -2x_1^4 + 8x_1^3 -8x_1^2 + x_2 - 2  \\
        -4x_1^4 + 32x_1^3 - 88x_1^2 + 96x_1 + x_2 -36
    \end{bmatrix}
$$

## The Constraint Jacobian Matrix

$$
    \nabla J(x) = \begin{bmatrix}
        -8x_1^3 + 24x_1 - 16 x_1 && 1 \\
        -16x_1^3 + 96x_1^2 - 176x_1 + 96 && 1
    \end{bmatrix}
$$

## The Constraint Hessian Matrix

$$
    \nabla H(x, v) = v_0 \begin{bmatrix}
        -24x_1^2 + 8 & 0 \\ 0 & 0
    \end{bmatrix} + v_1 \begin{bmatrix}
        -48x_1^2 + 192x_1 - 176 & 0 \\ 0 & 0
    \end{bmatrix}
$$